core/src/main/java/org/apache/calcite/rel/type/RelDataTypeImpl.java

```diff
@@ -217,6 +217,12 @@ public int getScale() {
     return SCALE_NOT_SPECIFIED;
   }
 
+  /**
+   * Gets the {@link SqlTypeName} of this type.
+   * Sub-classes must override the method to ensure the resulting value is non-nullable.
+   *
+   * @return SqlTypeName, never null
+   */
   public SqlTypeName getSqlTypeName() {
     return null;
```

Comment:
Why is this null? How about `SqlTypeName.ANY` ?

Reply:
The code was like that for ages, and this `return null` is never used in Calcite (it is always overridden).
Another option could be removing this implementation (the class is `abstract` anyway), however, that sounds like a backward-incompatible change, so I just kept the method as is.

Reply:
The code looks weird, the document says it never returns null but there is a default code returns null.

Reply:
Can we replace it with another `SqlTypeName` instead of null ?

Reply:
I'm afraid no any default value is suitable here, for example `SqlTypeName .ANY`, because each `SqlTypeName` has a specific semantic, there is no default, returns a default value will cause confusion in some cases.

Reply:
If no default value is suitable, maybe we should throw an UnSupportedOperationException here to make sure it is always overriden.

Reply:
I would love to `un-implement` the method, however, it might happen that there is third-party code that extends the class, and it keeps the default `return null` implementation. It might even be the case that the default `null` implementation works because the code never tries to call methods on the result of `getSqlTypeName()`.

That is why I'm inclined to keep the default `return null`, and deal with it later (e.g. announce the method would become abstract in 6 months).

Reply:
Just in case, there's `getSqlIdentifier`:

```java
  public SqlTypeName getSqlTypeName() {
    // The implementations must provide non-null value, however, we keep this for compatibility
    return castNonNull(null);
  }

  public @Nullable SqlIdentifier getSqlIdentifier() {
    SqlTypeName typeName = getSqlTypeName();
    if (typeName == null) {
      return null;
    }
```

So I would keep the method return null as before.